rfcs/2022-02-11-responsive-and-adaptive-layout-strategy.md

```diff
@@ -0,0 +1,394 @@
+| ------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------ |
+| [Paste - Twilio](https://paste.twilio.design/primitives/text/#responsive-text)                          | Flexible with `[]` props accepting values for breakpoints on various components. Uses styled system under the hood                   |
+| [Primer - GitHub](https://primer.style/react/core-concepts#responsive-props)                            | Flexible with `[]` props accepting props accepting values for breakpoints on various components. Uses styled system under the hood   |
+| [Arcade](https://arcade.design/content/docs/tokens/typography)                                          | Handles internally, not customisable on the consumer end(nice strategy for less overhead)                                            |
+| [React Spectrum - Adobe](https://react-spectrum.adobe.com/react-spectrum/layout.html#responsive-layout) | Flexible to define breakpoints for the entire application and flexible to define behavior for those breakpoints per components basis |
+
+## What would work for us?
+
+As you saw there are pros and cons to each of the approach so we can't abandon either of them but we need to strike a strategy where each of these approaches will work best.
+
+After doing some research and aligning it with our uses cases at Razorpay, I propose following things at a broader level.
+
+### 1. Components that can handle Responsiveness Internally
+
+- Most of the components that don't deal with layouts shall handle responsiveness internally. For eg: Button, Modal, Avatar, Select, etc.
+
+- The properties that affects responsiveness are:
+
+  - `width`
+  - `height`
+  - `font-size`
+  - `padding`
+  - `margin`
+  - `position`
+
+- While a lot of [well known design systems gives flexibility](#approaches-taken-by-well-known-design-systems) on every component for above properties to accept responsive values, I can't think of a use case where we'll have to provide similar functionality. Majority of the cases that I could think of wouldn't be any different even if the consumers define the values for them explicitly. So seems like we'll be adding unnecessary learning curve to the component's API.
+
+  - For eg: all the props(ones that are affected by responsive) for a component will then accept either single value or array/object of values for different screen sizes which is confusing and might lead to a lot of confusion at the API surface.
+
+    ```jsx
+    <Text fontSize="100" />
+
+    <Text fontSize={{ small: 100, medium: 100, large: 100, xl: 400 }} />
+    ```
+
+- Hence the idea here is that all these components except the Layout components will handle responsiveness based on predefined breakpoints internally without the consumers explicitly defining anything for them.
```

Comment:
@kamleshchandnani The components handling responsiveness on there own seem analogous to having design tokens, i.e. these promote consistency among all the consumer projects. If that is the case, then to adopt the Blade components, the respective designs will also need to be updated. How are we exposing these components along with the defined presents in figma?

Reply:
There's a slight difference WRT tokens vs components here. so the tokens will define the scale of spacing/margin/sizing for these components on different resolutions. Coming to design side they actually don't have to do much here because for them they'll always have to create multiple mockups because AFAIK we don't have that responsiveness in design tools yet like we do in browsers(except the auto layout).

On Figma, these components will have a property of screen size as a variant which the designers will toggle when placing the components in different screen frames

Reply:
Got it. So let's say they have different designs for different screens, how will the component dimensions look to the designers when they set a particular variant based on the screen size?

My question is, will setting the dimensions for these components in the design, be a manual process?

Reply:
okay, so 2 parts over here:

1. The design on Figma for Design System side
2. The designs for the consumer side on Figma

So for the first case, assume if the components have a width that varies then the designer on the design system side will publish components such that it has variable width(auto-layout) and then when the consumer side designer uses it on any screen it won't matter

Another scenario where the components are different on different screen sizes for eg: modal with max-width property which takes full width on mobile and 80% on a desktop the designer on the design system side will design 2 different elements(even though it's a responsive modal) since we don't have matured design tools yet :D and then the designer on the consuming side need to use the version of the Modal based on the screen size they pick. Unfortunately, it's a manual process because again the limitation on how the current design tools are built.

Reply:
@sidag95 Yes, it's a manual process but designers just need to change the property in Figma's right panel to change the design from large to small screen.

Reply:
Understood. Thanks for the detailed explanation.